compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs

```diff
@@ -3365,4 +3365,113 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             err.span_label(block.span, "this block is missing a tail expression");
         }
     }
+
+    pub(crate) fn annotate_incorrect_or_expr(
+        &self,
+        diag: &mut Diag<'_>,
+        expr: &'tcx hir::Expr<'tcx>,
+        expr_ty: Ty<'tcx>,
+        expected_ty: Ty<'tcx>,
+    ) {
+        if expected_ty != self.tcx.types.bool {
+            return;
+        }
+        let hir::Node::Expr(&hir::Expr {
+            kind:
+                hir::ExprKind::Binary(
+                    hir::BinOp { node: hir::BinOpKind::Or, span: binop_span },
```

Comment:
We could also allows this for `|` (just allowing mixing `|` and `||` to keep the diagnostic impl simple seems fine)

Reply:
That has different semantics (and hir structure compared to this check). The below works:

```rust
fn main() {
    let a = 10;
    if a == 1 | 2 | 3 {
        
    }
}
```